### PR TITLE
vault and simple voting implementations

### PR DESCRIPTION
**Pull request by novaknole**

### CODE REVIEW

File: lerna.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (nivida) [inline lerna.json:1]: EOF.. sorry haha ;-)

File: package.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-	"name": "zaragoza",
+	"name": "@aragon/zaragoza-monorepo",
```
Comment (nivida) [inline package.json:2]: @aragon/zaragoza and later on probably @aragon/aragon.

File: package.json [hunk 3]
```diff
@@ -9,5 +9,8 @@
 	"workspaces": [
 		"packages/*"
 	],
-	"scripts": {}
+	"scripts": {},
+	"devDependencies": {
+		"lerna": "^3.22.1"
```
Comment (nivida) [inline package.json:14]: Is lerna required if we have yarn workspaces?

File: package.json [hunk 4]
```diff
@@ -9,5 +9,8 @@
 	"workspaces": [
 		"packages/*"
 	],
-	"scripts": {}
+	"scripts": {},
```
Comment (nivida) [inline package.json:12]: empty `scripts`

File: packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol [hunk 5]
```diff
@@ -61,24 +70,26 @@ abstract contract GovernancePrimitive is UUPSUpgradeable, Initializable {
...
 
-        Execution memory execution = Execution({
-            process: process,
-            proposal: proposal,
-            state: GovernancePrimitive.State.RUNNING
-        });
+        executionsCounter++;
 
-        _start(execution); // "Hook" to add logic in start of a concrete implementation.
+        // the reason behind this - https://matrix.to/#/!poXqlbVpQfXKWGseLY:gitter.im/$6IhWbfjcTqmLoqAVMopWFuIhlQwsoaIRxmsXhhmsaSs?via=gitter.im&via=matrix.org&via=ekpyron.org
```
Comment (nivida) [inline packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol:83]: lol.. love this language..

File: packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol [hunk 6]
```diff
@@ -35,19 +35,28 @@ abstract contract GovernancePrimitive is UUPSUpgradeable, Initializable {
...
+    modifier executionExist(uint256 _id) {
+        require(_id < executionsCounter, ERROR_NO_EXECUTION);
+        _;
+    }
+
     /// @dev Used for UUPS upgradability pattern
     /// @param _dao The DAO contract of the current DAO
-    function initialize(DAO _dao) external initializer {
-            dao = _dao;
+    function initialize(DAO _dao) public initializer {
```
Comment (nivida) [inline packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol:58]: why public?

File: packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol [hunk 7]
```diff
@@ -61,24 +70,26 @@ abstract contract GovernancePrimitive is UUPSUpgradeable, Initializable {
     /// @param process The process definition.
     /// @param proposal The proposal for execution submitted by the user.
     /// @return executionId The id of the newly created execution.
-    function start(Processes.Process calldata process, Proposal calldata proposal) external returns (uint256 executionId) {
+    function start(Processes.Process calldata process, Proposal calldata proposal) public returns (uint256 executionId) {
```
Comment (nivida) [inline packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol:73]: Just as a reminder to me and also you.. we have to check the state in eg. execute and other methods.. but quite simple to add this require.

File: packages/contracts/contracts/src/processes/Processes.sol [hunk 8]
```diff
@@ -90,7 +90,7 @@ contract Processes is UUPSUpgradeable, Initializable {
                 if (action.to == allowedAction.to) { // CONTRACT MATCHED
                     uint256 methodsLength = allowedAction.methods.length;
                     for (uint256 y = 0; y < methodsLength; y++) { // CHECK FOR EVERY ALLOWD METHOD OF A CONTRACT
-                        if (action.signature == allowedAction.methods[y]) { // METHOD FOUND STOP SEARCHING
+                        if (bytes4(action.data[:4]) == allowedAction.methods[y]) { // METHOD FOUND STOP SEARCHING
```
Comment (nivida) [inline packages/contracts/contracts/src/processes/Processes.sol:93]: Thanks to 0.8.6 🥳

File: packages/contracts/package.json [hunk 9]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "zaragoza",
+  "name": "@aragon/zaragoza-contracts",
```
Comment (nivida) [inline packages/contracts/package.json:2]: Let us discuss about having a package called core for the core contracts and two for the concrete implementation of the governance primitives and permission validators..

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 10]
```diff
@@ -1,49 +1,40 @@
...
-        bytes data;
-    }
-
     enum VoterState { Absent, Yea, Nay }
 
     struct Vote {
-        bool executed;
-        uint64 startDate;
-        uint64 snapshotBlock;
+        uint256 voteId;
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:19]: I will think about this again also if we have discussed about.. hmm idk.. there has to be a cleaner way..

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 11]
```diff
@@ -1,49 +1,40 @@
...
+        uint256 startDate;
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:20]: why uint256 for the timestamp and snapshotBlock?

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 12]
```diff
@@ -62,11 +53,6 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
     event ChangeMinQuorum(uint64 minAcceptQuorumPct);
 
-    modifier voteExists(uint256 _voteId) {
-        require(_voteId < votesLength, ERROR_NO_VOTE);
-        _;
-    }
-
     // TODO: @Giorgi check inheritance cause of initialize
     function initialize(DAO _dao, ERC20VotesUpgradeable _token, uint64[3] calldata _voteSettings) external initializer { 
         dao = _dao;
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:58]: I think we can now remove line 72

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 13]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
-    //     Vote storage vote_ = votes[voteId];
-    //     vote_.startDate = block.timestamp; // TODO:
-    //     vote_.snapshotBlock = snapshotBlock;
-    //     vote_.supportRequiredPct = supportRequiredPct;
-    //     vote_.minAcceptQuorumPct = minAcceptQuorumPct;
-    //     vote_.votingPower = votingPower;
-    //     vote_.actions = proposal.actions;
+        Vote storage vote_ = votes[voteId];
+        vote_.voteId = voteId;
+        vote_.startDate = block.timestamp; // TODO:
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:114]: todo?

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 14]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
-        
-    //     // TODO: 
-    //     // 1. do we wanna add one more if which will decide if 
-    //     // the creator of the vote should vote or not immediatelly.
-    //     if (_canVote(voteId, msg.sender)) {
-    //         _vote(voteId, true, msg.sender, executeIfDecided);
-    //     }
-        
-    // }
+        emit StartVote(voteId, msg.sender, description);
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:120]: shouldn't we better store the description off-chain and just reference a hash?

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 15]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
+        }
+    }
 
-    //function start(Data.ProposalSubmission calldata proposal) public override {
-        // (Action[] memory actions, string memory description, bool executeIfDecided) = abi.decode(proposal.data, [Action[], string, bool]);
+    /**
+    * @dev Overriden function that actually gets called from the Process.
+    * @param data abi encoded data that includes necessary parameters to vote.
+    */
+    function _vote(uint256 _voteId, bytes calldata data) internal override {
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:131]: I think we can merge those two _vote methods.

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 16]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
+        if (castVote && canVote(voteId, msg.sender)) {
+            _vote(voteId, true, msg.sender, executeIfDecided);
+        }
+    }
 
-    //function start(Data.ProposalSubmission calldata proposal) public override {
-        // (Action[] memory actions, string memory description, bool executeIfDecided) = abi.decode(proposal.data, [Action[], string, bool]);
+    /**
+    * @dev Overriden function that actually gets called from the Process.
+    * @param data abi encoded data that includes necessary parameters to vote.
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:129]: voteId missing in natspec.. also we can add more information.. have a look at the other contracts.. the thing is if we define the nat-spec's well can we generate better code references etc.

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 17]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
-        // require(votingPower > 0, ERROR_NO_VOTING_POWER);
+        require(_canVote(_voteId, msg.sender), ERROR_CAN_NOT_VOTE);
+
+        _vote(_voteId, _supports, msg.sender, _executesIfDecided);
+    }
+
+    /**
+    * @dev Internal function to cast a vote. It assumes the queried vote exists.
+    */
+    function _vote(uint256 _voteId, bool _supports, address _voter, bool _executesIfDecided) internal {
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:145]: Params etc missing in natspec

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 18]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
 
-    //function canVote(uint256 _voteId, address _voter) public view voteExists(_voteId) returns (bool) {
-    //    return _canVote(_voteId, _voter);
-    //}
-/*
     function getVote(uint256 _voteId)
         public
         view
-        voteExists(_voteId)
+        executionExist(_voteId)
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:196]: Is this required to have it as modifier? Probably better to have it as internal method cause modifiers will get duplicated..

Comment (nivida) [reply packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:196]: We also can have this check directly in _getExecution and revert if it isn't existing..

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 19]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
-        if (_executesIfDecided && _canExecute(_voteId)) {
-           _executeVote(_voteId);
-        }
-    }
-/*
     /**
     * @dev Internal function to check if a voter can participate on a vote. It assumes the queried vote exists.
     * @return True if the given voter can participate a certain vote, false otherwise
-    *//*
+    */
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:227]: missing natspec

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 20]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
+
     /**
     * @dev Internal function to check if a vote is still open
     * @return True if the given vote is open, false otherwise
     */
-    /*
     function _isVoteOpen(Vote storage vote_) internal view returns (bool) {
-        return block.timestamp < vote_.startDate + voteTime && !vote_.executed; // TODO:
-    }*/
+        return block.timestamp < vote_.startDate + voteTime && _isVoteExecuted(vote_.voteId); // TODO:
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:238]: Todo?

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol [hunk 21]
```diff
@@ -78,231 +64,196 @@ contract SimpleVoting is VotingGovernancePrimitive {
...
+    }
+
+    /**
+    * @dev Internal function to check if a vote is executed
+    * @return True if the given vote is open, false otherwise
+    */
+    function _isVoteExecuted(uint256 _voteId) internal view returns(bool) {
```
Comment (nivida) [inline packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:245]: Let us add this state check to the abstract GovernancePrimitive and are there any reasons to have it here as well?

**Review by nivida — CHANGES_REQUESTED** [review]
First of all a big thank you for finishing the Vault and SimpleVoting. I've added a few comments and will have a second look at 1-2 methods of the SimpleVoting contract tomorrow morning.

File: packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol [hunk 22]
```diff
@@ -35,19 +35,28 @@ abstract contract GovernancePrimitive is UUPSUpgradeable, Initializable {
...
     }
 
-    uint256 public executionsCounter;
-    mapping(uint256 => Execution) public executions;
+    uint256 private executionsCounter;
+    mapping(uint256 => Execution) private executions;
+
     DAO internal dao;
 
+    string private constant ERROR_NO_EXECUTION = "ERROR_NO_EXECUTION";
```
Comment (nivida) [inline packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol:49]: This is not used right?

File: packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol (hunk 6 above)
Comment (novaknole) [reply packages/contracts/contracts/lib/governance-primitives/GovernancePrimitive.sol:58]: @nivida when it's external, you can't call it from child contract at all. So moved it to public. This shouldn't bring any problems as the contracts don't call it internally in other places. so we should be good.

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol (hunk 10 above)
Comment (novaknole) [reply packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:19]: can we merge first and think about this later ?

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol (hunk 11 above)
Comment (novaknole) [reply packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:20]: Initially, `block.timestamp` and startDate return `uint256`. In the couple of days, I plan on having `uint64`.

File: packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol (hunk 13 above)
Comment (novaknole) [reply packages/contracts/contracts/src/governance-primitives/voting/SimpleVoting.sol:114]: The `todo` means I have to transform it to uint64 later on.

**Pull request closed without merging**